README
   webcheck - a website link checker

   webcheck was originally named linbot which was developed by Albert
   Hopkins (marduk) <marduk@python.net>.

   Versions up till 1.0 were maintained by Mike W. Meyer <mwm@mired.org> who
   changed the name to webcheck, http://www.mired.org/webcheck/.

   After that Arthur de Jong <arthur@arthurdejong.org> took up the work and did
   a complete rewrite, http://arthurdejong.org/webcheck/.

   Copyright (C) 1998, 1999 Albert Hopkins (marduk)
   Copyright (C) 2002 Mike W. Meyer
   Copyright (C) 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2013 Arthur de Jong

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA

   The files produced as output from the software do not automatically fall
   under the copyright of the software, unless explicitly stated otherwise.

   webcheck includes the FancyTooltips javascript library to display readable
   tooltips. FancyTooltips is distributed under the MIT license and has the
   following copyright notices (see fancytooltips/fancytooltips.js for
   details):

   Copyright (C) 2005 Victor Kulinski
   Copyright (C) 2003 Dunstan Orchard, Ethan Marcotte, Mark Wubben
   Copyright (C) 2003 Stuart Langridge, Paul McLanahan, Peter Janes,
                 Brad Choate

INTRODUCTION
============

webcheck is a website checking tool for webmasters. It crawls a given website
and generates a number of reports. The whole system is pluggable allowing
easily adding extra reports and checks.

Features of webcheck include:
 * support for http, https, ftp and file schemes
 * view the structure of a site
 * track down broken links
 * find potentially outdated and new pages
 * list links pointing to external sites
 * can run without user intervention

webcheck is written in Python and is developed on a Debian system with
Python 2.7. Previous versions of Python are not tested regularly. Patches
to support a wider range of Python releases are welcome (provided they are
not too intrusive).


INSTALLING WEBCHECK
===================

This will install the latest version from PyPi.

  % easy_install webcheck


MANUAL INSTALLATION
===================

Installation is relatively easy. These installation instructions are for
Unix-like systems. Other operating systems may differ.

  1. Unpack the tarball in the location that you want to have it installed.
     Maybe something like /usr/local/lib/python/site-packages or /opt.
     % tar -xvzf webcheck-1.10.4.tar.gz

  2. Add a symbolic link to some place in your PATH.
     % ln -s /opt/webcheck-1.10.4/webcheck.py /usr/local/bin/webcheck

  3. Put the manual page in the MANPATH.
     % ln -s /opt/webcheck-1.10.4/webcheck.1 /usr/local/man/man1/webcheck.1


RUNNING WEBCHECK
================

Executing webcheck without any command line arguments will cause it to give a
simple synopsis of its usage and then quit. Giving it the --help option will
cause it to print out all command line options.

Webcheck writes its reports to an output directory. This is the current
directory by default. Running webcheck as something like:

  % webcheck -o /tmp/myreport http://www.example.com/

Should crawl the site and write the reports to the /tmp/myreport directory.
The reports are simple HMTL pages that should look fine with most modern
browsers.

For more information on webcheck usage and command line options, see the
webcheck manual page. If the manual page is not in the MANPATH you can
probably open the manual with something like:
  % man -l /opt/webcheck-1.10.4/webcheck.1


FEEDBACK AND BUG REPORTS
========================

If you have any questions about webcheck or would like to report a bug, please
include the following:

 * the exact webcheck command that was run
 * output of webcheck (prefferably with the --debug option)
 * gzipped webcheck.dat file from the output directory

(a simple HTML file demonstrating the bug would even be better)

Suggestions for improvements are also welcome. Patches and code contributions
are even better.

Please send all your reports to <webcheck-users@lists.arthurdejong.org>.


NEWS
changes from 1.10.4 to 1.10.5 (alpha)
-----------------------------

* added setup.py for pypi/egg-based installation
* support --max-depth option to control max depth
* detect and report on endless redirects
* move to sqlite for storing crawler state


changes from 1.10.3 to 1.10.4
-----------------------------

* remove some left-over debugging code
* several small bugfixes which more or less drop support for Python 2.3
* limit the list of "referenced from" to 10 items
* pass char_encoding option to tidy to fix some tidy-related errors
* add a Referer header if possible (thanks Devin Bayer)
* Debian packaging improvements


changes from 1.10.2 to 1.10.3
-----------------------------

* support <iframe> and some common usages of <object>
* fix bug in command-line parsing of short -r option
* implement the --userpass option to pass username and password information to
  specific sites based on a patch by Chris Shenton
* handle errors while parsing more gracefully
* add parsing of <script> tag and background attributes, based on a patch by
  Robert M. Jansen
* fix in parsing <style> tags and support style attributes
* call tidy (if available) on HTML content, based on a patch by Henning
  Sielaff
* fix problem with port numbers in host headers


changes from 1.10.1 to 1.10.2
-----------------------------

* add checking for bug in BeautifulSoup and issue warning if bug is found
* added support for Python 2.3 (alhough more recent versions of Python
  are recommended)
* small documentation improvements
* Debian package improvements


changes from 1.10.0 to 1.10.1
-----------------------------

* some extra Unicode handling precautions
* fix problem in reading webcheck.dat for non-ASCII text
* be more verbose about HTTP retrieval failures
* split out URL normalization code into own module and do some basic protocol-
  specific normalizations
* a number of big performance improvements
* fix a bug in handling some zero-size pages
* parse http-equiv meta HTML header to parse refresh option
* webcheck now requires Python 2.4 or more recent


changes from 1.9.8 to 1.10.0
----------------------------

* switched HTML parsing to using BeautifulSoup with a fall-back mechanism to
  the old HTMLParser based solution
* the new parser is much more error-tolerant but is reportedly somewhat slower
  and does not include line numbers in errors
* new features will likely only be added to the new parser
* some small improvements to the output to make it XHTML 1.1 compliant
* internal improvements for handling Unicode strings
* better support for parsing <applet> tags and anchors using id attributes
* re-enable robots.txt parsing that was disabled in 1.9.8 and add an
  --ignore-robots option


changes from 1.9.7 to 1.9.8
---------------------------

* some checks for properly handling unknown and wrong encodings have been
  added
* added proper error handling for SSL related socket problems (exceptions are
  not a subclass of regular socket exceptions)
* a bug fix for URLs that contain a user name without a password or the other
  way around
* miscellaneous small report improvements


changes from 1.9.6 to 1.9.7
---------------------------

* site data is now stored to a file while crawling the site, this can be used
  to resume a crawl with the --continue option and for debugging purposes
* implemented checking of link anchors
* small improvements to generated reports (favicon included, CSS fix)
* documentation improvements
* properly handle float values for --wait
* unreachable sites will time out faster
* added support for plugins that don't output html
* half a dozen other small bug fixes (stability fixes, code clean-ups and
  improvements)


changes from 1.9.5 to 1.9.6
---------------------------

* SECURITY FIX: a cross-site scripting vulnerability with content in the
                tooltips of generated report was fixed by properly escaping
                all output (CVE-2006-1321)
* URLs are now URL encoded into a consistent form, solving some problems with
  URLs with non-ASCII characters
* no longer remove unreferenced redirects
* more debugging info in debug mode
* more fixes for escaping in generated reports and more support for sites in
  different character sets


changes from 1.9.4 to 1.9.5
---------------------------

* about page now has some more useful information
* proxy authentication is implemented
* fix for using relative paths as output directory
* add support for parsing html documents in different encodings
* ensure that all generated html output is properly escaped
* implemented --internal option to flag internal URLs with regular expressions
* documentation improvements
* several bug fixes to get webcheck more robust
* included FancyTooltips by Victor Kulinski to have nicer tooltips
* generated reports now have friendlier messages for when there  is nothing to
  report
* there is a Debian package


changes from 1.9.3 to 1.9.4
---------------------------

* split problems into link problems (errors retrieving the document) and page
  problems (parsing errors, wrong links, etc)
* some fixes and improvements to the layout of the generated pages
* redirect loops are now detected
* transfer result status is now stored
* addition of a limited CSS parser that handles imports and url() entries
* support reading file names for checking from the command line (turning them
  into file:// URLs internally)
* better error handling of problems writing generated pages and check that we
  are not overwriting input files


changes from 1.9.2 to 1.9.3
---------------------------

* several improvements to the generated reports, including tooltips with some
  useful information for the links (does not seem to work very well in
  Firefox)
* stability improvements to the html parser (thanks to everyone who reported
  problems) not all problems have been solved but it shouldn't stop webcheck
  any more
* reimplementation of the file and ftp modules to read directory contents or

webcheck.py
#!/usr/bin/env python

# webcheck.py - main module of webcheck doing command-line checking
#
# Copyright (C) 1998, 1999 Albert Hopkins (marduk)
# Copyright (C) 2002 Mike W. Meyer
# Copyright (C) 2005, 2006, 2007, 2008, 2010, 2011 Arthur de Jong
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
#
# The files produced as output from the software do not automatically fall
# under the copyright of the software, unless explicitly stated otherwise.

"""This is the main webcheck module."""

__version__ = '1.10.4'
__homepage__ = 'http://arthurdejong.org/webcheck/'

import sys
import os
import re
import urlparse
import urllib

import config
# update some fields that currently are stored in config
config.VERSION = __version__
config.HOMEPAGE = __homepage__

import crawler
import plugins
import debugio
import monkeypatch
import db

debugio.loglevel = debugio.INFO


def print_version():
    """Print version information."""
    sys.stdout.write(
      'webcheck %(version)s\n'
      'Written by Albert Hopkins (marduk), Mike W. Meyer and Arthur de Jong.\n'
      '\n'
      'Copyright (C) 1998-2011\n'
      'Albert Hopkins (marduk), Mike W. Meyer and Arthur de Jong.\n'
      'This is free software; see the source for copying conditions.  There is NO\n'
      'warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n'
      % {'version': __version__})


def print_usage():
    """Print short usage information."""
    sys.stderr.write(
      'Usage: webcheck [OPTION]... URL...\n')


def print_tryhelp():
    """Print friendly pointer to more information."""
    sys.stderr.write(
      'Try \'webcheck --help\' for more information.\n')


def print_help():
    """Print the option list."""
    sys.stdout.write(
      'Usage: webcheck [OPTION]... URL...\n'
      'Generate a report for the given URLs\n'
      '\n'
      '  -i, --internal=PATTERN mark URLs matching PATTERN as internal\n'
      '  -x, --external=PATTERN mark URLs matching PATTERN as external\n'
      '  -y, --yank=PATTERN     do not check URLs matching PATTERN\n'
      '  -b, --base-only        base URLs only: consider any URL not starting\n'
      '                         with any of the base URLs to be external\n'
      '  -a, --avoid-external   do not check external URLs\n'
      '      --ignore-robots    do not retrieve and parse robots.txt files\n'
      '  -q, --quiet, --silent  suppress progress messages\n'
      '  -d, --debug            do programmer-level debugging\n'
      '  -o, --output=DIRECTORY store the generated reports in the specified\n'
      '                         directory\n'
      '  -c, --continue         try to continue from a previous run\n'
      '  -f, --force            overwrite files without asking\n'
      '  -r, --redirects=N      the number of redirects webcheck should follow,\n'
      '                         0 implies to follow all redirects (default=%(redirects)d)\n'
      '  -u, --userpass=URL     specify a URL with user:pass so username and password are given\n'
      '                         to matching network locations, -u http://user:pass@example.com\n'
      '  -w, --wait=SECONDS     wait SECONDS between retrievals\n'
      '  -V, --version          output version information and exit\n'
      '  -h, --help             display this help and exit\n'
      % {'redirects': config.REDIRECT_DEPTH})


def parse_args(site):
    """Parse command-line arguments."""
    import getopt
    try:
        optlist, args = getopt.gnu_getopt(sys.argv[1:],
          'i:x:y:l:baqdo:cfr:u:w:Vh',
          ('internal=', 'external=', 'yank=', 'base-only', 'avoid-external',
           'ignore-robots',
           'quiet', 'silent', 'debug', 'profile', 'output=', 'continue',
           'force', 'redirects=', 'userpass=', 'wait=', 'version', 'help'))
        internal_urls = []
        external_urls = []
        yank_urls = []
        for flag, arg in optlist:
            if flag in ('-i', '--internal'):
                internal_urls.append(arg)
            elif flag in ('-x', '--external'):
                external_urls.append(arg)
            elif flag in ('-y', '--yank'):
                yank_urls.append(arg)
            elif flag in ('-b', '--base-only'):
                config.BASE_URLS_ONLY = True
            elif flag in ('-a', '--avoid-external'):
                config.AVOID_EXTERNAL_LINKS = True
            elif flag in ('--ignore-robots',):
                config.USE_ROBOTS = False
            elif flag in ('-q', '--quiet', '--silent'):
                debugio.loglevel = debugio.ERROR
            elif flag in ('-d', '--debug'):
                debugio.loglevel = debugio.DEBUG
            elif flag in ('--profile',):
                # undocumented on purpose
                config.PROFILE = True
            elif flag in ('-o', '--output'):
                config.OUTPUT_DIR = arg
            elif flag in ('-c', '--continue'):
                config.CONTINUE = True
            elif flag in ('-f', '--force'):
                config.OVERWRITE_FILES = True
            elif flag in ('-r', '--redirects'):
                config.REDIRECT_DEPTH = int(arg)
            elif flag in ('-u', '--userpass'):
                (_scheme, _netloc, _path, _params, _query, _frag) = urlparse.urlparse(arg)
                (_userpass, _netloc) = urllib.splituser(_netloc)
                config.USERPASS[_netloc] = _userpass
            elif flag in ('-w', '--wait'):
                config.WAIT_BETWEEN_REQUESTS = float(arg)
            elif flag in ('-V', '--version'):
                print_version()
                sys.exit(0)
            elif flag in ('-h', '--help'):
                print_help()
                sys.exit(0)
        if len(args) == 0 and not config.CONTINUE:
            print_usage()
            print_tryhelp()
            sys.exit(1)
        # ensure output directory exists
        if not os.path.isdir(config.OUTPUT_DIR):
            os.mkdir(config.OUTPUT_DIR)
        # set up database connection
        filename = os.path.join(config.OUTPUT_DIR, 'webcheck.sqlite')
        from sqlalchemy import create_engine
        engine = create_engine('sqlite:///' + filename)
        db.Session.configure(bind=engine)
        # ensure that all tables are created
        db.Base.metadata.create_all(engine)
        # TODO: schema migraton goes here
        # add configuration to site
        for pattern in internal_urls:
            site.add_internal_re(pattern)
        for pattern in external_urls:
            site.add_external_re(pattern)
        for pattern in yank_urls:
            site.add_yanked_re(pattern)
        for arg in args:
            # if it does not look like a url it is probably a local file
            if urlparse.urlsplit(arg)[0] == '':
                arg = 'file://' + urllib.pathname2url(os.path.abspath(arg))
            site.add_internal(arg)
    except getopt.error, reason:
        sys.stderr.write('webcheck: %s\n' % reason)
        print_tryhelp()
        sys.exit(1)
    except re.error, e:
        sys.stderr.write('webcheck: %s\n' % str(e))
        sys.exit(1)


def install_file(source, text=False):
    """Install the given file in the output directory.
    If the text flag is set to true it is assumed the file is text,
    translating line endings."""
    import shutil
    import urlparse
    # figure out mode to open the file with
    mode = 'r'
    if text:
        mode += 'U'
    # check with what kind of argument we are called
    scheme = urlparse.urlsplit(source)[0]
    if scheme == 'file':
        # this is a file:/// url, translate to normal path and open
        import urllib
        source = urllib.url2pathname(urlparse.urlsplit(source)[2])
    elif scheme == '' and os.path.isabs(source):
        # this is an absolute path, just open it as is
        pass
    elif scheme == '':
        # this is a relavite path, try to fetch it from the python path
        for directory in sys.path:
            tst = os.path.join(directory, source)
            if os.path.isfile(tst):
                source = tst
                break
    # TODO: support more schemes here
    # figure out the destination name
    target = os.path.join(config.OUTPUT_DIR, os.path.basename(source))
    # test if source and target are the same
    source = os.path.realpath(source)
    if source == os.path.realpath(target):
        debugio.warn('attempt to overwrite %(fname)s with itself' % {'fname': source})
        return
    # open the input file
    sfp = None
    try:
        sfp = open(source, mode)
    except IOError, (errno, strerror):
        debugio.error('%(fname)s: %(strerror)s' %
                      {'fname': source,
                       'strerror': strerror})
        sys.exit(1)
    # create file in output directory (with overwrite question)
    tfp = plugins.open_file(os.path.basename(source))
    # copy contents
    shutil.copyfileobj(sfp, tfp)
    # close files
    tfp.close()
    sfp.close()


def main(site):
    """Main program."""
    # crawl through the website
    debugio.info('checking site....')
    crawler.setup_urllib2()
    site.crawl()  # this will take a while
    debugio.info('done.')
    # do postprocessing (building site structure, etc)
    debugio.info('postprocessing....')
    site.postprocess()
    debugio.info('done.')
    # now we can write out the files
    # start with the frame-description page
    debugio.info('generating reports...')
    # for every plugin, generate a page
    site.generate()
    # put extra files in the output directory
    install_file('webcheck.css', True)
    install_file('fancytooltips/fancytooltips.js', True)
    install_file('favicon.ico', False)
    debugio.info('done.')


if __name__ == '__main__':
    try:
        # initialize site object
        site = crawler.Site()
        # parse command-line arguments
        parse_args(site)
        # run the main program
        if config.PROFILE:
            fname = os.path.join(config.OUTPUT_DIR, 'webcheck.prof')
            try:
                import cProfile
            except ImportError:
                import profile as cProfile
            try:
                import sqltap
                sqltap.start()
            except ImportError:
                pass
            cProfile.run('main(site)', fname)
            if 'sqltap' in locals():
                statistics = sqltap.collect()
                sqltap.report(statistics, os.path.join(config.OUTPUT_DIR, 'sqltap.html'))
        else:
            main(site)
    except KeyboardInterrupt:
        sys.stderr.write('Interrupted\n')
        sys.exit(1)
